ee/trie_store: refactor LazyTrieLeaf into LazilyDeserializedTrie

alsrdn opened the pull request:
This provides stronger type checking and allows us to remove useless paths in the code that are unreachable.

Fixes: https://github.com/casper-network/casper-node/issues/3994

alsrdn committed:
ee/trie_store: refactor LazyTrieLeaf into LazilyDeserializedTrie

This provides stronger type checking and allows us to remove useless
paths in the code that are unreachable.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store/operations: use try_deserialize_leaf_key

Use LazilyDeserializedTrie::try_deserialize_leaf_key for the `delete`
operation instead of deserializing the key manually.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
node/Cargo.toml: update openssl to 0.10.55

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented on execution_engine/src/storage/trie/tests.rs:
```diff
@@ -92,12 +92,73 @@ mod pointer_block {
+
+    use crate::storage::trie::{gens::*, LazilyDeserializedTrie, PointerBlock, Trie};
+
+    fn test_trie_roundtrip_to_lazy_trie<K, V>(trie: &Trie<K, V>, check_key: bool)
```
Do we need the `check_key` param here? Or maybe we can infer it below?

`let check_key = matches!(deserialized_from_slice, LazilyDeserializedTrie::Leaf(_))`?

alsrdn replied:
Fixed in: https://github.com/casper-network/casper-node/pull/4083/commits/765677f7c07be92bd21cbfd3303dfcf9e74cd819

rafal-ch reviewed (approved):
LGTM :+1: 
No blockers. Good stuff :)

mpapierski commented on execution_engine/src/storage/trie/mod.rs:
```diff
@@ -511,40 +510,97 @@ impl<K, V> Trie<K, V> {
+                }
+            }
+            LazilyDeserializedTrie::Node { .. } | LazilyDeserializedTrie::Extension { .. } => {
+                Err(bytesrepr::Error::Formatting)
```
I'm thinking the `Formatting` error may be confusing here. Technically, the "Trie" deserialization was successful, so what do you think about this:

```suggestion
                Ok(None)
```

alsrdn replied:
Yes, makes more sense to return Ok(None) here, though we'll have to convert the error in the callers somehow. I'll try to make the change.

alsrdn replied:
Went into a slightly different direction since returning and Optional value here was leading to a lot of useless code. We are actually getting the key only from a leaf so we shouldn't even be ever looking at the other variants. Added a new type for the leaf bytes in this commit: https://github.com/casper-network/casper-node/pull/4083/commits/765677f7c07be92bd21cbfd3303dfcf9e74cd819

mpapierski commented on execution_engine/src/storage/trie/mod.rs:
```diff
@@ -511,40 +510,97 @@ impl<K, V> Trie<K, V> {
+            LazilyDeserializedTrie::Leaf(leaf_bytes) => {
+                let (tag_byte, rem) = u8::from_bytes(leaf_bytes)?;
+                let tag = TrieTag::from_u8(tag_byte).ok_or(bytesrepr::Error::Formatting)?;
+                if let TrieTag::Leaf = tag {
```
This could be an `assert!` as well, no?

alsrdn replied:
Hmm I'm not sure. If the bytes are not those of a leaf it's probably a deserialization error, right? I believe that when we deserialize a key and value for `Trie<K, V>` we would return `Error::Formatting`. WDYT?

alsrdn replied:
Added the assert in https://github.com/casper-network/casper-node/pull/4083/commits/765677f7c07be92bd21cbfd3303dfcf9e74cd819

mpapierski commented on execution_engine/src/storage/trie_store/operations/tests/scan.rs:
```diff
@@ -50,16 +57,22 @@ where
     }
-    assert_eq!(root, tip.expect_right("Unexpected leaf found"));
+
+    assert_matches!(
+        tip,
+        LazilyDeserializedTrie::Node { .. } | LazilyDeserializedTrie::Extension { .. },
+        "Unexpected leaf found"
+    );
+    assert_eq!(root, tip.try_into()?);
```
```suggestion
  assert!(matches!(
    tip,
    LazilyDeserializedTrie::Node { .. } | LazilyDeserializedTrie::Extension { .. }),
    "Unexpected leaf found"));
    assert_eq!(root, tip.try_into()?);
```

alsrdn replied:
Fixed in https://github.com/casper-network/casper-node/pull/4083/commits/765677f7c07be92bd21cbfd3303dfcf9e74cd819

mpapierski commented on node/Cargo.toml:
```diff
@@ -56,7 +56,7 @@ num-rational = { version = "0.4.0", features = [ "serde" ] }
 num_cpus = "1"
 once_cell = "1"
-openssl = "0.10.32"
+openssl = "0.10.55"
```
Is this necessary in this PR 🤔

alsrdn replied:
I'm OK in having it as a separate PR. I only put it here because we need to update the dependency for the CI to pass.

alsrdn replied:
Reverted this change for now. We can open a new PR with this or just merge back `dev` to this branch since the change is already there. WDYT?

rafal-ch replied:
I think we should merge `dev` to `feat-1.6`.

alsrdn committed:
Revert "node/Cargo.toml: update openssl to 0.10.55"

This reverts commit 795bb487b9389a6524cf060d1af84eb23c48f852.

alsrdn committed:
ee/trie_store: address code review comments

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

mpapierski commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -1100,7 +1065,7 @@ where
                                 match self.store.get_raw(self.txn, pointer.hash()) {
                                     Ok(Some(trie_bytes)) => {
-                                        match trie::lazy_trie_deserialize(trie_bytes) {
+                                        match bytesrepr::deserialize(trie_bytes.into()) {
```
```suggestion
                                        match bytesrepr::deserialize_from_slice(&trie_bytes) {
```

alsrdn replied:
fixed in https://github.com/casper-network/casper-node/pull/4083/commits/42c4d61a9a2fea9d299d65b128d3854d869f1870

mpapierski commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -1150,7 +1115,8 @@ where
                     if affix.starts_with(&check_prefix) {
                         maybe_next_trie = match self.store.get_raw(self.txn, pointer.hash()) {
-                            Ok(Some(trie_bytes)) => match trie::lazy_trie_deserialize(trie_bytes) {
+                            Ok(Some(trie_bytes)) => match bytesrepr::deserialize(trie_bytes.into())
```
```suggestion
                            Ok(Some(trie_bytes)) => match bytesrepr::deserialize_from_slice(&trie_bytes)
```

alsrdn replied:
fixed in https://github.com/casper-network/casper-node/pull/4083/commits/42c4d61a9a2fea9d299d65b128d3854d869f1870

mpapierski commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -1206,10 +1169,10 @@ where
         Ok(None) => (vec![], KeysIteratorState::Ok),
         Err(e) => (vec![], KeysIteratorState::ReturnError(e)),
-        Ok(Some(current_root_bytes)) => match trie::lazy_trie_deserialize(current_root_bytes) {
+        Ok(Some(current_root_bytes)) => match bytesrepr::deserialize(current_root_bytes.into()) {
```
```suggestion
        Ok(Some(current_root_bytes)) => match bytesrepr::deserialize_from_slice(&current_root_bytes) {
```

alsrdn replied:
fixed in https://github.com/casper-network/casper-node/pull/4083/commits/42c4d61a9a2fea9d299d65b128d3854d869f1870

mpapierski commented on execution_engine/src/storage/trie_store/lmdb.rs:
```diff
@@ -219,9 +219,9 @@ impl ScratchTrieStore {
-                trie::lazy_trie_deserialize(trie_bytes.clone())?;
-            tries_to_write.extend(trie::lazy_trie_iter_children(&lazy_trie));
+            let lazy_trie: LazilyDeserializedTrie =
+                bytesrepr::deserialize(trie_bytes.clone().into())?;
```
Not sure if this will compile but worth a try:

```suggestion
                bytesrepr::deserialize_from_slice(&trie_bytes)?;
```

alsrdn replied:
fixed in https://github.com/casper-network/casper-node/pull/4083/commits/42c4d61a9a2fea9d299d65b128d3854d869f1870

mpapierski commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -325,24 +325,17 @@ where
-                // since we are checking if this is a leaf and skipping, we do not expect to ever
-                // hit this.
-                unreachable!()
+        let maybe_trie_leaf = bytesrepr::deserialize(current.into())?;
```
```suggestion
        let maybe_trie_leaf = bytesrepr::deserialize_from_slice(&current)?;
```

mpapierski reviewed (approved):
one nit regarding using `bytesrepr::deserialize` where `bytesrepr::deserialize_from_slice` could be used, otherwise LGTM

alsrdn committed:
ee/trie_store: use bytesrepr::deserialize_from_slice when appropriate

Use `bytesrepr::deserialize_from_slice` instead of
`bytesrepr::deserialize` when it's more appropriate to do so to avoid
needless conversion.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
Merge branch 'feat-1.6' into LazyTrieLeaf-refactor

alsrdn committed:
ee/trie_store: fix formatting issues

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>
